File: CartQuestiOS/CartQuestiOS/Repositories/CartRepository.swift
import Foundation
import FirebaseAuth
import FirebaseFirestore

class CartRepository {
    private lazy var db = Firestore.firestore()

    private func cartsCollection() -> CollectionReference {
        guard let uid = Auth.auth().currentUser?.uid else {
            fatalError("User must be authenticated to access carts")
        }
        return db.collection("users").document(uid).collection("carts")
    }

    func getActiveCart() async throws -> Cart? {
        let snapshot = try await cartsCollection()
            .whereField("status", isEqualTo: "active")
            .limit(to: 1)
            .getDocuments()

        guard let doc = snapshot.documents.first else { return nil }
        var cart = try doc.data(as: Cart.self)
        cart.id = doc.documentID
        return cart
    }

    func saveCart(_ cart: Cart) async throws -> Cart {
        var cartToSave = cart
        cartToSave.updatedAt = Date()

        let docRef: DocumentReference
        if cart.id.isEmpty {
            docRef = cartsCollection().document()
            cartToSave.id = docRef.documentID
        } else {
            docRef = cartsCollection().document(cart.id)
        }

        try docRef.setData(from: cartToSave)
        return cartToSave
    }

    func completeCart(cartId: String) async throws {
        try await cartsCollection().document(cartId).updateData(["status": "completed"])
    }
}

File: CartQuestAndroid/app/src/main/java/com/amoghghadge/cartquestandroid/data/repository/CartRepository.kt
package com.amoghghadge.cartquestandroid.data.repository

import com.amoghghadge.cartquestandroid.data.model.Cart
import com.google.firebase.auth.ktx.auth
import com.google.firebase.firestore.ktx.firestore
import com.google.firebase.firestore.ktx.toObject
import com.google.firebase.ktx.Firebase
import kotlinx.coroutines.tasks.await

class CartRepository {
    private val db = Firebase.firestore
    private val auth = Firebase.auth

    private fun cartsCollection() =
        db.collection("users").document(auth.currentUser!!.uid).collection("carts")

    suspend fun getActiveCart(): Cart? {
        val snapshot = cartsCollection()
            .whereEqualTo("status", "active")
            .limit(1)
            .get().await()
        return snapshot.documents.firstOrNull()?.let { doc ->
            doc.toObject<Cart>()?.copy(id = doc.id)
        }
    }

    suspend fun saveCart(cart: Cart): Cart {
        val docRef = if (cart.id.isEmpty()) cartsCollection().document() else cartsCollection().document(cart.id)
        val savedCart = cart.copy(id = docRef.id, updatedAt = System.currentTimeMillis())
        docRef.set(savedCart).await()
        return savedCart
    }

    suspend fun completeCart(cartId: String) {
        cartsCollection().document(cartId).update("status", "completed").await()
    }
}
